// depcheck/dist/special/webpack.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function concat(array, item) {
  return array.concat(item);
}

function duplicate(array, item) {
  return array.indexOf(item) === -1 ? array.concat([item]) : array;
}

function extractLoaders(item) {
  if (item.loader) {
    return item.loader.split('!');
  } else if (item.loaders) {
    return item.loaders;
  }

  return [];
}

function stripQueryParameter(loader) {
  var index = loader.indexOf('?');
  return index === -1 ? loader : loader.substring(0, index);
}

function normalizeLoader(deps, loader) {
  var templates = ['*-webpack-loader', '*-web-loader', '*-loader', '*'];
  var names = templates.map(function (template) {
    return template.replace('*', loader);
  }).filter(function (name) {
    return deps.indexOf(name) !== -1;
  });

  return names[0];
}

function getLoaders(deps, loaders) {
  return (loaders || []).map(extractLoaders).reduce(concat, []).map(function (loader) {
    return stripQueryParameter(loader);
  }).map(function (loader) {
    return normalizeLoader(deps, loader);
  }).filter(function (loader) {
    return loader;
  }).reduce(duplicate, []);
}

exports.default = function (content, filepath, deps) {
  var filename = _path2.default.basename(filepath);
  if (filename === 'webpack.config.js') {
    var module = require(filepath).module || {};
    var loaders = getLoaders(deps, module.loaders);
    var preLoaders = getLoaders(deps, module.preLoaders);
    var postLoaders = getLoaders(deps, module.postLoaders);
    return loaders.concat(preLoaders).concat(postLoaders);
  }

  return [];
};

module.exports = exports['default'];
